QraphQL errors in browser

leimonio opened the pull request:
This approach to displaying GraphQL compilation and runtime errors in the browser console utilizes the WebSocket connecting node running gatsby package responsible for executing GraphQL queries and the cache loaded and used on the browser side (thanks @pieh for helping in this).

One more useful step would be to extract the error query and add a link to graphiql with the error query.

KyleAMathews commented:
Woah cool!!!

Would love to see a screenshot!

How will errors get cleared? What's your thinking there?

leimonio commented:
Screenshot of the error:
![screen shot 2018-07-02 at 09 52 55](https://user-images.githubusercontent.com/1969742/42154751-7b257238-7dde-11e8-92c3-3167ab9d42b9.png)

Good catch regarding clearing errors. Not sure yet, what's the strategy, should the page get reloaded reload or HMR should be fine? Right now when you edit query to a non-erroneous one and HMR happens, it doesn't show the compiled page.

KyleAMathews commented:
Oh nice! Looks great!

On clearing errors — I was under the impression this would show up in the Error Overlay but I guess not?

leimonio commented:
Not sure, what it's actually blocking error from displaying in the Error Overlay. I've tried 2 ways, `console.error(new Error(msg.payload.error))` and `ErrorOverlay.reportBuildError(new Error(msg.payload.error))` but nothing of them did work. I'm wondering where the issue might be. Maybe this is because `socketIo.js` loads before `root.js`? Do you have any idea I should try?

leimonio commented:
Ok, managed to make it appear on the Error Overlay, so now it appears both on the console and on screen. Needs some more fine-tuning with clearing errors on reload.

<img width="689" alt="screen shot 2018-07-02 at 23 00 25" src="https://user-images.githubusercontent.com/1969742/42188459-e90194f4-7e4b-11e8-8bb2-3ae1ed464ef8.png">

pieh commented:
I just tried using `ErrorOverlay.reportBuildError("test")` and it worked. Using `ErrorOverlay.reportBuildError(new Error("test"))` didn't work, because overlay seems to expect only strings (got tons of errors in console). 

But to use overlay we need to figure out
- clearing errors when they are fixed - I think we could accumulate errors when we are compiling and running queries and emit array of errors when we are done - if error array is not empty -> show overlay and hide it if it's empty
- playing nice with build errors - i.e. if we have both syntax errors in js and errors in graphql, so fixing just one doesn't hide overlay - I think we need to wrap calls to `reportBuildError`/`dismissBuildError` in something that will keep track of errors from different sources, so it knows to hide overlay only there are no build and graphql errors

leimonio commented:
@pieh Really good points raised, glad you went through with review and feedback on the 
edge cases. 💯 

I've updated the PR, solving these points. Aggregated error handling utilizing `ErrorOverlay` 
under a single utility module, also added some tests. PTAL and let me know what do you think.

leimonio commented:
@pieh @KyleAMathews Is there anything further you'd recommend for this PR?

pieh commented:
I need some time to battle test it locally - hope to do that tomorrow!

calcsam commented:
@leimonio sorry this got merge conflicts! If you're able to resolve those @Chuloo should be able to do some battle testing and then we can get this merged in!

leimonio commented:
@calcsam I will take a look in the conflicts one of these days :) thanks for getting back!

leimonio commented:
@calcsam @Chuloo is there a way to manually trigger rebuild on travis? Dropdown doesn't display any such option on failing tests.

pieh commented:
@leimonio I did some changes - will need to adjust test for overlay

I was testing it locally and we can't rely on emiting query results to clear overlay errors - there are too many edge cases there plus soon there we will change how gatsby develop is running queries (it will run queries on demand instead of always prerunning all of them). So for now it will just show graphql compiling errors and not runtime errors.

pieh commented:
and just preview of what it does for other folks:
![v28xkdwnfb](https://user-images.githubusercontent.com/419821/46804073-076eef80-cd62-11e8-8e5b-169b94ce5e9e.gif)

leimonio commented:
@pieh what are the alternatives to emitting query results for runtime errors? Is there anything in place I could utilize? Otherwise, I can remove handling of runtime errors and proceed with that.

pieh commented:
> @pieh what are the alternatives to emitting query results for runtime errors? Is there anything in place I could utilize? Otherwise, we can proceed with your last update.

Emitting errors is/was correct - just clearing error part is problematic - we would want to clear those when query for path that previously was problematic finished succesfuly, but there are edge cases - like what happens if we delete page/or just query all together without fixing up query etc - that's why I decided we shouldn't try to fix that for initial rollout. Also runtime graphql errors are far less common than compile errors, so this still provide good value

pieh commented:
Sorry for taking a long time to merge this @leimonio! `gatsby@2.0.44` was published with those changes

leimonio commented:
@pieh No worries, there was a big release in the meantime and so many great news for GatsbyJS. Thanks for the support and helping out in the contribution process. Please feel free and let me know if there is any other issue I can further help!
